compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1648,8 +1665,32 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
 
   def typedExpr(tree: untpd.Tree, pt: Type = WildcardType)(implicit ctx: Context): Tree =
     typed(tree, pt)(ctx retractMode Mode.PatternOrType)
-  def typedType(tree: untpd.Tree, pt: Type = WildcardType)(implicit ctx: Context): Tree = // todo: retract mode between Type and Pattern?
+  def typedType(tree: untpd.Tree, pt: Type = WildcardType)(implicit ctx: Context): Tree = {
+    // todo: retract mode between Type and Pattern?
+
+    /** Check that the are not mixed Any/Phantom.Any types in `&`, `|` and type bounds,
+     *  this includes Phantom.Any of different universes.
+     */
+    def checkedTops(tree: untpd.Tree): Set[Type] = {
+      def checkedTops2(tree1: untpd.Tree, tree2: untpd.Tree, msg: => Message, pos: Position): Set[Type] = {
+        val allTops = checkedTops(tree1) union checkedTops(tree2)
+        if (allTops.size > 1)
+          ctx.error(msg, tree.pos)
+        allTops
+      }
+      tree match {
+        case TypeBoundsTree(lo, hi) =>
+          checkedTops2(lo, hi, PhantomCrossedMixedBounds(lo, hi), tree.pos)
+        case untpd.InfixOp(left, op, right) =>
+          checkedTops2(left, right, PhantomCrossedMixedBounds(left, right), tree.pos)
+        case EmptyTree => Set.empty
+        case _ => Set(defn.topOf(tree.typeOpt))
+      }
+    }
+    checkedTops(tree)
```

Comment:
I think it's better to reorganize type checking as follows: In TypeAssigner, insert check code in the assignType methods for AndType, OrType, TypeBounds, If, and Match. Here's an example for OrType:

    def assignType(tree: untpd.OrTypeTree, left: Tree, right: Tree)(implicit ctx: Context) = {
      checkSameUniverse(left, right, "be combined in `|`)
      tree.withType(left.tpe | right.tpe)
    }

where

    checkSameUniverse(tree1: Tree, tree2: Tree, relationship: => String) =
      if (topClass(tree1.tpe) != topClass(tree2.tpe)) 
        ctx.error(ex"${tree1.tpe} and ${tree2.tpe} are in different universes. They cannot $relationship")

If we do this, `Typer` can be left completely alone.